wordpress-mobile/WordPress-Android
# Train of likes animation

Fixes #15251

Internal Ref: `pctCYC-8w-p2`
iOS main PR: https://github.com/wordpress-mobile/WordPress-iOS/pull/16936

Due to AFKs and Sabbaticals pinging folks outside of my team, thanks for any support you can give; trying to give a bit more context in the description below 🙇 .

The following is the final result (avatar duplication intended to not show other users avatar)

https://user-images.githubusercontent.com/47797566/130979978-b7c6357d-07e5-41b4-b024-afb0dfaf66ee.mov

The original implementation (without the animation) had the "n bloggers like this" text outside the recycler; since we need a synched animation it seemed fair to bring the text inside the recycler and leverage the fade in/out recycler built in animation.

Files involved are a few in number but the changes were pretty related and hopefully are not too heavy to review 🤞 (in GH at least for me it rendered better using the split diff view); just adding a bit of grouping in the hope that it helps:
- Files under `android.ui.reader` package are the main modification. They:
    - Move the "n bloggers like this" as an item inside the recycler
    - Decouple `ReaderPostDetailViewModel` and `EngagedPeopleListViewModel` introducing a dedicated TrainOfFacesUiState + relevant Diff Callback and View Holders
    - Manage the like/unlike operation locally in the reader details screen, without querying the endpoint
- 4 files (those under `android.ui.engagement` package) are changes that remove unused fields from `EngagedPeopleListUiState` + the "limit and retry" logic that is no more needed given we use pageLength for the limit and manage the like/unlike operation locally without requesting the likers data from the endpoint
- remaining 8 files are unit testing and small layout changes.

A last note about the "n bloggers like this"; given the `You` is the current user, possible state we expect are:

- <ins>You</ins> like this: current user is the only liker
- <ins>You and 1 blogger</ins> like this
- <ins>You and n bloggers</ins> like this
- <ins> 1 blogger</ins> likes this
- <ins>n blogger</ins> like this

## To test
1. Borrowing below the test steps from the iOS linked PR:
    - Open a post with 0 likes. 
      - Verify the likes view is shown when the post is liked.
      - Verify the likes view is hidden when the post is unliked.
    - Open a post where you're the only one liking the post.
      - Verify the likes view is hidden when the post is unliked.
      - Verify the likes view is shown when the post is liked.
    - Open a post with more than 5 likes from other people. 
      - Verify that it shows 6 avatars when the post is liked.
      - Verify that it shows 5 avatars when the post is unliked.
    - Verify that the animation works fine when the post is liked and unliked.
    - Verify that the animation direction is correct in RTL language.

2. Smoke test likers list by tapping on the train of faces and accessing from Notifications > Likes

## Regression Notes
1. Potential unintended areas of impact
List of likes in ReaderPostDetails and Notifications > Likes

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing + already existing unit testing

3. What automated tests I added (or what prevented me from doing so)
Added a few unit tests where applicable

PR submission checklist:
